### PR TITLE
Add ClusterDistribution field for CNS telemetry

### PR DESCRIPTION
**Pull request by yuyiying**
This pull request is adding a go binding for ClusterDistribution to support CNS telemetry and also make cns/client.go backward compatible to vsan67u3 and vsan70 by dropping optional fields not known to the prior releases.

Testing:
For VC version > 7.0
```
% export CNS_VC_URL='https://Administrator@vsphere.local:Admin!23@10.78.100.14/sdk'
% export CNS_DATACENTER='VSAN-DC'
% export CNS_DATASTORE='vsanDatastore'

% go test -v
=== RUN   TestClient
--- PASS: TestClient (32.61s)
    client_test.go:128: Creating volume using the spec: types.CnsVolumeCreateSpec{
            Name:       "pvc-901e87eb-c2bd-11e9-806f-005056a0c9a0",
            VolumeType: "BLOCK",
            Datastores: []types.ManagedObjectReference{
                {Type:"Datastore", Value:"datastore-36"},
            },
            Metadata: types.CnsVolumeMetadata{
                ContainerCluster: types.CnsContainerCluster{
                    ClusterType:         "KUBERNETES",
                    ClusterId:           "demo-cluster-id",
                    VSphereUser:         "Administrator@vsphere.local",
                    ClusterFlavor:       "VANILLA",
                    ClusterDistribution: "OpenShift",
                },
                EntityMetadata:        nil,
                ContainerClusterArray: nil,
            },
            BackingObjectDetails: &types.CnsBlockBackingDetails{
                CnsBackingObjectDetails: types.CnsBackingObjectDetails{
                    CapacityInMb: 5120,
                },
                BackingDiskId:      "",
                BackingDiskUrlPath: "",
            },
            Profile:    nil,
            CreateSpec: nil,
        }
.
.
.
client_test.go:217: Successfully Queried Volumes. queryResult: &types.CnsQueryResult{
            Volumes: []types.CnsVolume{
                {
                    VolumeId: types.CnsVolumeId{
                        Id: "0cdfb822-2ddd-4897-a067-3fe37682371d",
                    },
                    DatastoreUrl:    "ds:///vmfs/volumes/vsan:524f0c83003d697c-b67ae4bec09d8333/",
                    Name:            "pvc-901e87eb-c2bd-11e9-806f-005056a0c9a0",
                    VolumeType:      "BLOCK",
                    StoragePolicyId: "aa6d5a82-1c88-45da-85d3-3d74b91a5bad",
                    Metadata:        types.CnsVolumeMetadata{
                        ContainerCluster: types.CnsContainerCluster{
                            ClusterType:         "KUBERNETES",
                            ClusterId:           "demo-cluster-id",
                            VSphereUser:         "Administrator@vsphere.local",
                            ClusterFlavor:       "VANILLA",
                            ClusterDistribution: "OpenShift",
                        },
                        EntityMetadata:        nil,
                        ContainerClusterArray: []types.CnsContainerCluster{
                            {
                                ClusterType:         "KUBERNETES",
                                ClusterId:           "demo-cluster-id",
                                VSphereUser:         "Administrator@vsphere.local",
                                ClusterFlavor:       "VANILLA",
                                ClusterDistribution: "OpenShift",
                            },
                        },
                    },
.
.
.
PASS
ok  	github.com/vmware/govmomi/cns	32.642s
```
simulator test
```
=== RUN   TestSimulator
--- PASS: TestSimulator (0.04s)
PASS
ok  	github.com/vmware/govmomi/cns/simulator	0.064s
```
For VC version <= 7.0
Before modifying the dropping unknown fields:
```
% export CNS_VC_URL='https://Administrator@vsphere.local:Admin!23@10.182.2.135/sdk'
% export CNS_DATACENTER='VSAN-DC'
% export CNS_DATASTORE='vsanDatastore'
% go test -v

    client_test.go:131: Failed to create volume. Error: ServerFaultCode:
        Unexpected element tag "clusterDistribution" seen

   client_test.go:380: Failed to update volume metadata. Error: ServerFaultCode:
        Unexpected element tag "clusterDistribution" seen

```
After modify:
```
=== RUN   TestClient
--- PASS: TestClient (46.75s)
    client_test.go:128: Creating volume using the spec: types.CnsVolumeCreateSpec{
            Name:       "pvc-901e87eb-c2bd-11e9-806f-005056a0c9a0",
            VolumeType: "BLOCK",
            Datastores: []types.ManagedObjectReference{
                {Type:"Datastore", Value:"datastore-36"},
            },
            Metadata: types.CnsVolumeMetadata{
                ContainerCluster: types.CnsContainerCluster{
                    ClusterType:         "KUBERNETES",
                    ClusterId:           "demo-cluster-id",
                    VSphereUser:         "Administrator@vsphere.local",
                    ClusterFlavor:       "VANILLA",
                    ClusterDistribution: "OpenShift",
                },
                EntityMetadata:        nil,
                ContainerClusterArray: nil,
            },
            BackingObjectDetails: &types.CnsBlockBackingDetails{
                CnsBackingObjectDetails: types.CnsBackingObjectDetails{
                    CapacityInMb: 5120,
                },
                BackingDiskId:      "",
                BackingDiskUrlPath: "",
            },
            Profile:    nil,
            CreateSpec: nil,
        }
    client_test.go:153: Volume created sucessfully. volumeId: f5a9e13a-e2f7-4621-a416-60eda374f23f
.
.
.
client_test.go:211: Calling QueryVolume using queryFilter: types.CnsQueryFilter{
            VolumeIds: []types.CnsVolumeId{
                {
                    Id: "f5a9e13a-e2f7-4621-a416-60eda374f23f",
                },
            },
            Names:                        nil,
            ContainerClusterIds:          nil,
            StoragePolicyId:              "",
            Datastores:                   nil,
            Labels:                       nil,
            ComplianceStatus:             "",
            DatastoreAccessibilityStatus: "",
            Cursor:                       (*types.CnsCursor)(nil),
            healthStatus:                 "",
        }
    client_test.go:217: Successfully Queried Volumes. queryResult: &types.CnsQueryResult{
            Volumes: []types.CnsVolume{
                {
                    VolumeId: types.CnsVolumeId{
                        Id: "f5a9e13a-e2f7-4621-a416-60eda374f23f",
                    },
                    DatastoreUrl:    "ds:///vmfs/volumes/vsan:52216ea322833634-e5928c73810e5a3d/",
                    Name:            "pvc-901e87eb-c2bd-11e9-806f-005056a0c9a0",
                    VolumeType:      "BLOCK",
                    StoragePolicyId: "aa6d5a82-1c88-45da-85d3-3d74b91a5bad",
                    Metadata:        types.CnsVolumeMetadata{
                        ContainerCluster: types.CnsContainerCluster{
                            ClusterType:         "KUBERNETES",
                            ClusterId:           "demo-cluster-id",
                            VSphereUser:         "Administrator@vsphere.local",
                            ClusterFlavor:       "VANILLA",
                            ClusterDistribution: "",
                        },
.
.
.
    client_test.go:376: Updating volume using the spec: {DynamicData:{} VolumeId:{DynamicData:{} Id:f5a9e13a-e2f7-4621-a416-60eda374f23f} Metadata:{DynamicData:{} ContainerCluster:{DynamicData:{} ClusterType:KUBERNETES ClusterId:demo-cluster-id VSphereUser:Administrator@vsphere.local ClusterFlavor:VANILLA ClusterDistribution:OpenShift} EntityMetadata:[0xc00040ea00 0xc00040ea80 0xc00040eb00] ContainerClusterArray:[{DynamicData:{} ClusterType:KUBERNETES ClusterId:demo-cluster-id VSphereUser:Administrator@vsphere.local ClusterFlavor:VANILLA ClusterDistribution:OpenShift}]}}
    client_test.go:401: Successfully updated volume metadata
client_test.go:410: Successfully Queried Volumes. queryResult: &types.CnsQueryResult{
            Volumes: []types.CnsVolume{
                {
                    VolumeId: types.CnsVolumeId{
                        Id: "f5a9e13a-e2f7-4621-a416-60eda374f23f",
                    },
                    DatastoreUrl:    "ds:///vmfs/volumes/vsan:52216ea322833634-e5928c73810e5a3d/",
                    Name:            "pvc-901e87eb-c2bd-11e9-806f-005056a0c9a0",
                    VolumeType:      "BLOCK",
                    StoragePolicyId: "aa6d5a82-1c88-45da-85d3-3d74b91a5bad",
                    Metadata:        types.CnsVolumeMetadata{
                        ContainerCluster: types.CnsContainerCluster{
                            ClusterType:         "KUBERNETES",
                            ClusterId:           "demo-cluster-id",
                            VSphereUser:         "Administrator@vsphere.local",
                            ClusterFlavor:       "VANILLA",
                            ClusterDistribution: "",
                        },
.
.
.
PASS
ok  	github.com/vmware/govmomi/cns	46.777s

```

@SandeepPissay @divyenpatel

### CODE REVIEW

**Comment by divyenpatel** [comment]
@yuyiying Can you make this change backward compatible.

Refer how we are dropping unknown fields for prior releases of vSphere. - https://github.com/vmware/govmomi/blob/master/cns/cns_util.go#L71-L105

**Review by divyenpatel — CHANGES_REQUESTED** [review]
Please address https://github.com/vmware/govmomi/pull/2113#issuecomment-692226968

**Comment by SandeepPissay** [comment]
> @yuyiying Can you make this change backward compatible.
> 
> Refer how we are dropping unknown fields for prior releases of vSphere. - https://github.com/vmware/govmomi/blob/master/cns/cns_util.go#L71-L105

@divyenpatel Why is this required? Does older CNS fail an API request that has newer fields?

**Comment by divyenpatel** [comment]
Yes @SandeepPissay, if we do not drop unknown elements, call with fail on the older release with the message `Unexpected element tag seen while parsing serialized DataObject`. 

With this [code](https://github.com/vmware/govmomi/blob/master/cns/cns_util.go#L71-L105), we are making sure the latest CSI driver works on older releases of vSphere.

**Comment by SandeepPissay** [comment]
@divyenpatel This looks like a problem in govmomi. Govmomi should ignore those fields that the server does not understand. IIRC this is how other vmomi clients behave. Cc @dougm

**Comment by divyenpatel** [comment]
@yuyiying Can you squash all your commits. Change looks good to me.

**Comment by yuyiying** [comment]
@divyenpatel  Already squashed. Thanks!

File: cns/cns_util.go [hunk 1]
```diff
@@ -106,11 +108,12 @@ func dropUnknownCreateSpecElements(c *Client, createSpecList []cnstypes.CnsVolum
 
 // dropUnknownVolumeMetadataUpdateSpecElements helps drop newly added elements in the CnsVolumeMetadataUpdateSpec, which are not known to the prior vSphere releases
 func dropUnknownVolumeMetadataUpdateSpecElements(c *Client, updateSpecList []cnstypes.CnsVolumeMetadataUpdateSpec) []cnstypes.CnsVolumeMetadataUpdateSpec {
+	var updatedUpdateSpecList []cnstypes.CnsVolumeMetadataUpdateSpec
```
Comment (SandeepPissay) [inline cns/cns_util.go:111]: Can we initialize this array with the length of updateSpecList?

File: cns/cns_util.go [hunk 2]
```diff
@@ -123,6 +126,18 @@ func dropUnknownVolumeMetadataUpdateSpecElements(c *Client, updateSpecList []cns
...
+			updateSpec.Metadata.ContainerCluster.ClusterDistribution = ""
+			var updatedContainerClusterArray []cnstypes.CnsContainerCluster
+			for _, containerCluster := range updateSpec.Metadata.ContainerClusterArray {
+				containerCluster.ClusterDistribution = ""
+				updatedContainerClusterArray = append(updatedContainerClusterArray, containerCluster)
+			}
+			updateSpec.Metadata.ContainerClusterArray = updatedContainerClusterArray
+			updatedUpdateSpecList = append(updatedUpdateSpecList, updateSpec)
+		}
+		updateSpecList = updatedUpdateSpecList
```
Comment (SandeepPissay) [inline cns/cns_util.go:140]: Looks like we are updating the param and also returning it. Why is this required?

File: cns/cns_util.go (hunk 1 above)
Comment (yuyiying) [reply cns/cns_util.go:111]: Done

File: cns/cns_util.go (hunk 2 above)
Comment (yuyiying) [reply cns/cns_util.go:140]: We need this parameter because when the version is not VSAN67U3 nor VSAN70, then we wont change it.

**Review by divyenpatel — APPROVED** [review]

**Comment by divyenpatel** [comment]
@dougm Can you help merge this PR?

**Review by dougm — APPROVED** [review]
Thanks @yuyiying
